File: frontend/recetario/lib/data/models/portafolio.dart
/// Modelo de Portafolio desde el backend
class Portafolio {
  final String id;
  final String estudianteId;
  final String titulo;
  final String? descripcion;
  final String ingredientes;
  final String preparacion;
  final List<String> fotos;
  final String? videoUrl;
  final String categoriaId;
  final String tipoReceta; // 'propia' o 'api'
  final String? fuenteApiId;
  final String visibilidad; // 'publica' o 'privada'
  final String? nivelAlcanzado;
  final int likes;
  final int vistas;
  final bool esDestacada;
  final bool esCertificada;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Campos adicionales cuando viene con info del estudiante
  final String? nombreEstudiante;
  final String? avatarEstudiante;
  final String? codigoEstudiante;

  Portafolio({
    required this.id,
    required this.estudianteId,
    required this.titulo,
    this.descripcion,
    required this.ingredientes,
    required this.preparacion,
    required this.fotos,
    this.videoUrl,
    required this.categoriaId,
    required this.tipoReceta,
    this.fuenteApiId,
    required this.visibilidad,
    this.nivelAlcanzado,
    required this.likes,
    required this.vistas,
    required this.esDestacada,
    required this.esCertificada,
    required this.createdAt,
    required this.updatedAt,
    this.nombreEstudiante,
    this.avatarEstudiante,
    this.codigoEstudiante,
  });

  factory Portafolio.fromJson(Map<String, dynamic> json) {
    return Portafolio(
      id: json['id'] ?? '',
      estudianteId: json['estudiante_id'] ?? '',
      titulo: json['titulo'] ?? '',
      descripcion: json['descripcion'],
      ingredientes: json['ingredientes'] ?? '',
      preparacion: json['preparacion'] ?? '',
      fotos: json['fotos'] != null 
          ? List<String>.from(json['fotos'])
          : [],
      videoUrl: json['video_url'],
      categoriaId: json['categoria_id'] ?? '',
      tipoReceta: json['tipo_receta'] ?? 'propia',
      fuenteApiId: json['fuente_api_id'],
      visibilidad: json['visibilidad'] ?? 'publica',
      nivelAlcanzado: json['nivel_alcanzado'],
      likes: json['likes'] ?? 0,
      vistas: json['vistas'] ?? 0,
      esDestacada: json['es_destacada'] ?? false,
      esCertificada: json['es_certificada'] ?? false,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
      nombreEstudiante: json['nombre_estudiante'],
      avatarEstudiante: json['avatar_estudiante'],
      codigoEstudiante: json['codigo_estudiante'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'estudiante_id': estudianteId,
      'titulo': titulo,
      'descripcion': descripcion,
      'ingredientes': ingredientes,
      'preparacion': preparacion,
      'fotos': fotos,
      'video_url': videoUrl,
      'categoria_id': categoriaId,
      'tipo_receta': tipoReceta,
      'fuente_api_id': fuenteApiId,
      'visibilidad': visibilidad,
      'nivel_alcanzado': nivelAlcanzado,
      'likes': likes,
      'vistas': vistas,
      'es_destacada': esDestacada,
      'es_certificada': esCertificada,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'nombre_estudiante': nombreEstudiante,
      'avatar_estudiante': avatarEstudiante,
      'codigo_estudiante': codigoEstudiante,
    };
  }

  Portafolio copyWith({
    String? id,
    String? estudianteId,
    String? titulo,
    String? descripcion,
    String? ingredientes,
    String? preparacion,
    List<String>? fotos,
    String? videoUrl,
    String? categoriaId,
    String? tipoReceta,
    String? fuenteApiId,
    String? visibilidad,
    String? nivelAlcanzado,
    int? likes,
    int? vistas,
    bool? esDestacada,
    bool? esCertificada,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? nombreEstudiante,
    String? avatarEstudiante,
    String? codigoEstudiante,
  }) {
    return Portafolio(
      id: id ?? this.id,
      estudianteId: estudianteId ?? this.estudianteId,
      titulo: titulo ?? this.titulo,
      descripcion: descripcion ?? this.descripcion,
      ingredientes: ingredientes ?? this.ingredientes,
      preparacion: preparacion ?? this.preparacion,
      fotos: fotos ?? this.fotos,
      videoUrl: videoUrl ?? this.videoUrl,
      categoriaId: categoriaId ?? this.categoriaId,
      tipoReceta: tipoReceta ?? this.tipoReceta,
      fuenteApiId: fuenteApiId ?? this.fuenteApiId,
      visibilidad: visibilidad ?? this.visibilidad,
      nivelAlcanzado: nivelAlcanzado ?? this.nivelAlcanzado,
      likes: likes ?? this.likes,
      vistas: vistas ?? this.vistas,
      esDestacada: esDestacada ?? this.esDestacada,
      esCertificada: esCertificada ?? this.esCertificada,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      nombreEstudiante: nombreEstudiante ?? this.nombreEstudiante,
      avatarEstudiante: avatarEstudiante ?? this.avatarEstudiante,
      codigoEstudiante: codigoEstudiante ?? this.codigoEstudiante,
    );
  }
}

/// Request para crear receta
class CrearPortafolioRequest {
  final String titulo;
  final String? descripcion;
  final String ingredientes;
  final String preparacion;
  final List<String> fotos;
  final String? videoUrl;
  final String categoriaId;
  final String tipoReceta;
  final String? fuenteApiId;
  final String visibilidad;

  CrearPortafolioRequest({
    required this.titulo,
    this.descripcion,
    required this.ingredientes,
    required this.preparacion,
    required this.fotos,
    this.videoUrl,
    required this.categoriaId,
    required this.tipoReceta,
    this.fuenteApiId,
    this.visibilidad = 'publica',
  });

  Map<String, dynamic> toJson() {
    return {
      'titulo': titulo,
      'descripcion': descripcion,
      'ingredientes': ingredientes,
      'preparacion': preparacion,
      'fotos': fotos,
      'video_url': videoUrl,
      'categoria_id': categoriaId,
      'tipo_receta': tipoReceta,
      'fuente_api_id': fuenteApiId,
      'visibilidad': visibilidad,
    };
  }
}

/// Modelo de Comentario desde backend
class ComentarioPortafolio {
  final String id;
  final String portafolioId;
  final String usuarioId;
  final String comentario;
  final DateTime createdAt;
  
  // Campos adicionales cuando viene con info del usuario
  final String? nombreUsuario;
  final String? avatarUsuario;

  ComentarioPortafolio({
    required this.id,
    required this.portafolioId,
    required this.usuarioId,
    required this.comentario,
    required this.createdAt,
    this.nombreUsuario,
    this.avatarUsuario,
  });

  factory ComentarioPortafolio.fromJson(Map<String, dynamic> json) {
    return ComentarioPortafolio(
      id: json['id'] ?? '',
      portafolioId: json['portafolio_id'] ?? '',
      usuarioId: json['usuario_id'] ?? '',
      comentario: json['comentario'] ?? '',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      nombreUsuario: json['nombre_usuario'],
      avatarUsuario: json['avatar_usuario'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'portafolio_id': portafolioId,
      'usuario_id': usuarioId,
      'comentario': comentario,
      'created_at': createdAt.toIso8601String(),
      'nombre_usuario': nombreUsuario,
      'avatar_usuario': avatarUsuario,
    };
  }
}

/// Modelo de Categoría
class Categoria {
  final String id;
  final String nombre;
  final String? descripcion;
  final String? icono;
  final int orden;
  final bool activo;

  Categoria({
    required this.id,
    required this.nombre,
    this.descripcion,
    this.icono,
    required this.orden,
    required this.activo,
  });

  factory Categoria.fromJson(Map<String, dynamic> json) {
    return Categoria(
      id: json['id'] ?? '',
      nombre: json['nombre'] ?? '',
      descripcion: json['descripcion'],
      icono: json['icono'],
      orden: json['orden'] ?? 0,
      activo: json['activo'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nombre': nombre,
      'descripcion': descripcion,
      'icono': icono,
      'orden': orden,
      'activo': activo,
    };
  }
}

File: backend/internal/models/portafolio.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type Portafolio struct {
	ID             uuid.UUID `json:"id"`
	UsuarioID      uuid.UUID `json:"usuario_id"` // ✅ CORREGIDO: era estudiante_id
	Titulo         string    `json:"titulo"`
	Descripcion    *string   `json:"descripcion,omitempty"`
	Ingredientes   string    `json:"ingredientes"`
	Preparacion    string    `json:"preparacion"`
	Fotos          []string  `json:"fotos"`
	VideoURL       *string   `json:"video_url,omitempty"`
	CategoriaID    uuid.UUID `json:"categoria_id"`
	TipoReceta     string    `json:"tipo_receta"` // 'propia' o 'api'
	FuenteAPIID    *string   `json:"fuente_api_id,omitempty"`
	Visibilidad    string    `json:"visibilidad"` // 'publica', 'privada'
	NivelAlcanzado *string   `json:"nivel_alcanzado,omitempty"`
	Likes          int       `json:"likes"`
	Vistas         int       `json:"vistas"`
	EsDestacada    bool      `json:"es_destacada"`
	EsCertificada  bool      `json:"es_certificada"`
	CreatedAt      time.Time `json:"created_at,omitempty"`
	UpdatedAt      time.Time `json:"updated_at,omitempty"`
}

// Portafolio con información del estudiante
type PortafolioConEstudiante struct {
	Portafolio
	NombreEstudiante string  `json:"nombre_estudiante"`
	AvatarEstudiante *string `json:"avatar_estudiante,omitempty"`
	CodigoEstudiante string  `json:"codigo_estudiante"`
}

// Request para crear receta propia
type CrearPortafolioRequest struct {
	Titulo       string   `json:"titulo" binding:"required,min=3,max=200"`
	Descripcion  *string  `json:"descripcion"`
	Ingredientes string   `json:"ingredientes" binding:"required,min=10"`
	Preparacion  string   `json:"preparacion" binding:"required,min=10"`
	Fotos        []string `json:"fotos" binding:"required,min=1"`
	VideoURL     *string  `json:"video_url"`
	CategoriaID  string   `json:"categoria_id" binding:"required,uuid"`
	TipoReceta   string   `json:"tipo_receta" binding:"required,oneof=propia api"`
	FuenteAPIID  *string  `json:"fuente_api_id"`
	Visibilidad  string   `json:"visibilidad" binding:"required,oneof=publica privada"`
}

// Request para actualizar receta
type ActualizarPortafolioRequest struct {
	Titulo       *string  `json:"titulo"`
	Descripcion  *string  `json:"descripcion"`
	Ingredientes *string  `json:"ingredientes"`
	Preparacion  *string  `json:"preparacion"`
	Fotos        []string `json:"fotos"`
	VideoURL     *string  `json:"video_url"`
	CategoriaID  *string  `json:"categoria_id"`
	Visibilidad  *string  `json:"visibilidad"`
}

// Comentario de portafolio
type ComentarioPortafolio struct {
	ID           uuid.UUID `json:"id"`
	PortafolioID uuid.UUID `json:"portafolio_id"`
	UsuarioID    uuid.UUID `json:"usuario_id"`
	Comentario   string    `json:"comentario"`
	CreatedAt    time.Time `json:"created_at,omitempty"`
}

// Comentario con información del usuario
type ComentarioConUsuario struct {
	ID            uuid.UUID `json:"id"`
	PortafolioID  uuid.UUID `json:"portafolio_id"`
	UsuarioID     uuid.UUID `json:"usuario_id"`
	Comentario    string    `json:"comentario"`
	CreatedAt     time.Time `json:"created_at,omitempty"`
	NombreUsuario string    `json:"nombre_usuario"`
	AvatarUsuario *string   `json:"avatar_usuario,omitempty"`
}

// ✅ UnmarshalJSON custom para mapear el JSON anidado de Supabase
func (c *ComentarioConUsuario) UnmarshalJSON(data []byte) error {
	// Primero parsear la estructura base
	type Alias ComentarioConUsuario
	aux := &struct {
		Usuarios *struct {
			NombreCompleto string  `json:"nombre_completo"`
			AvatarURL      *string `json:"avatar_url"`
		} `json:"usuarios"`
		*Alias
	}{
		Alias: (*Alias)(c),
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	// Mapear datos del usuario si existen
	if aux.Usuarios != nil {
		c.NombreUsuario = aux.Usuarios.NombreCompleto
		c.AvatarUsuario = aux.Usuarios.AvatarURL
	}

	return nil
}

// Request para crear comentario
type CrearComentarioRequest struct {
	Comentario string `json:"comentario" binding:"required,min=1,max=500"`
}

// Like de portafolio
type LikePortafolio struct {
	ID           uuid.UUID `json:"id"`
	PortafolioID uuid.UUID `json:"portafolio_id"`
	UsuarioID    uuid.UUID `json:"usuario_id"`
	CreatedAt    time.Time `json:"created_at,omitempty"`
}
